common/src/collision.rs

```diff
@@ -0,0 +1,588 @@
+use crate::node::DualGraph;
+use crate::{
+    dodeca::{Side, Vertex, SIDE_COUNT, VERTEX_COUNT},
+    graph::NodeId,
+};
+use lazy_static::lazy_static;
+use std::fmt;
+
+#[allow(dead_code)]
+/*
+The set of voxels that a collision body covers within a chunk
+*/
+#[derive(PartialEq, Clone, Copy)]
+pub struct ChunkBoundingBox {
+    pub node: NodeId,
+    pub chunk: Vertex,
+    pub min_xyz: na::Vector3<u32>,
+    pub max_xyz: na::Vector3<u32>,
+    pub dimension: u8,
+}
+
+pub struct VoxelAddress {
+    pub node: NodeId,
+    pub chunk: Vertex,
+    pub index: u32,
+}
+
+/*
+The set of voxels that a collision body covers.
+*/
+pub struct BoundingBox {
+    pub bounding_boxes: Vec<ChunkBoundingBox>,
+}
+
+// from a node coordinate in an arbitrary node, returns the which chunk the point would reside in
+pub fn chunk_from_location(location: na::Vector4<f64>) -> Option<Vertex> {
+    for v in Vertex::iter() {
+        let pos = (v.chunk_to_node().try_inverse().unwrap() * location).xyz();
+        if (pos.x >= 0_f64)
+            && (pos.x <= 1_f64)
+            && (pos.y >= 0_f64)
+            && (pos.y <= 1_f64)
+            && (pos.z >= 0_f64)
+            && (pos.z <= 1_f64)
+        {
+            return Some(v);
+        }
+    }
+    None
+}
+
+// Does not figure out which chunk it is in automatically
+// The pattern I use to unwrap the results of neighbor is kind of awkward.
+impl BoundingBox {
+    pub fn create_aabb(
+        start_node: NodeId,
+        position: na::Vector4<f64>,
+        radius: f64,
+        graph: &DualGraph,
+        dimension: u8,
+    ) -> Self {
+        assert!(
+            radius <= 1.0,
+            "Error: the radius of a bounding box may not exceed 1 absolute unit."
+        );
+
+        let start_chunk =
+            chunk_from_location(position).expect("Error: cannot find chunk for given position.");
+        let mut bounding_boxes = Vec::<ChunkBoundingBox>::new();
+        let sides = start_chunk.canonical_sides();
+
+        // get BBs for the chunks within the node.
+        for v in Vertex::iter() {
+            Self::add_sub_bb(
+                &mut bounding_boxes,
+                ChunkBoundingBox::get_chunk_bounding_box(
+                    start_node, v, position, radius, dimension,
+                ),
+            );
+        }
+
+        // get BBs for foreign chunks sharing a dodeca side
+        for side in sides.iter() {
+            if let Some(node) = graph.neighbor(start_node, *side) {
```

Comment:
The data inside the chunks don't need to be generated, the nodes themselves just have to exist. Is there a reason not to strictly enforce that as a precondition?

Reply:
The only reason I can think of is that it is likely possible to avoid enforcing this, which would reduce the probability of a crash. I'm not set either way.

Reply:
I think I would be in favor of using `unwrap` or `expect` here instead of `if let`, since if this issue does come up, it would then cause a crash instead of a subtle hard-to-debug issue.